paddle/fluid/operators/fused/mkldnn/multi_gru_mkldnn_op.cc

```diff
@@ -0,0 +1,694 @@
+    // Create memory key without Ti because weights, bias and h0 memories
+    // do not depend on Ti size but primitive and input/output memory do
+    if (platform::MKLDNNDeviceContext::tls().get_cur_mkldnn_session_id() !=
+        platform::MKLDNNDeviceContextThreadLocals::kMKLDNNSessionID_Default) {
+      memory_key_ = CreateKey(unique_name, MKLDNNGetDataType<T>());
+    } else {
+      memory_key_ = CreateKey(unique_name, MKLDNNGetDataType<T>(), "-t:",
+                              platform::ThreadIDasStr());
+    }
+    key_ = memory_key_;
+    key_.append("T").append(std::to_string(Ti_));
+
+    // Is it int8 kernel
+    const bool is_int8 = std::is_same<T, uint8_t>::value;
+
+    // Create attributes for each oneDNN gru
+    for (int i = 0; i < 2 * layers_; ++i) {
+      attrs_.push_back(dnnl::primitive_attr());
+    }
+
+    if (is_int8) {
+      // Add int8 attributes
+      const auto scale_weights = ctx.MultiInput<LoDTensor>("Scale_weights");
+      PADDLE_ENFORCE_EQ(
+          scale_weights.size(), layers_ * 2,
+          platform::errors::InvalidArgument(
+              "The number of weight scale inputs does "
+              "not match the number of layers. Expected: %d. Actual: %d",
+              layers_ * 2, scale_weights.size()));
+      const float scale_data = ctx.Attr<float>("Scale_data");
+      const float shift_data = ctx.Attr<float>("Shift_data");
+
+      const int weights_scale_mask =
+          0 +
+          (1 << 3)  // bit, indicating the unique scales for `g` dim in `ldigo`
+          +
+          (1 << 4);  // bit, indicating the unique scales for `o` dim in `ldigo`
+
+      int w_scale_num = scale_weights.size();
+      for (int i = 0; i < w_scale_num; ++i) {
+        attrs_[i].set_rnn_data_qparams(scale_data, shift_data);
+        const auto scale_weights_data = std::vector<float>(
+            scale_weights[i]->data<float>(),
+            scale_weights[i]->data<float>() + scale_weights[i]->numel());
+        attrs_[i].set_rnn_weights_qparams(weights_scale_mask,
+                                          scale_weights_data);
+      }
+    }
+
+    for (int layer = 0; layer < layers_; ++layer) {
+      AcquireGruPrimitiveDescriptor(layer, L2R);
+      AcquireGruPrimitiveDescriptor(layer, R2L);
+      AcquireConcatPrimitiveDescriptor(layer);
+    }
+  }
+
+  void AcquireGruPrimitiveDescriptor(int layer, Direction dir) {
+    auto pd_key = key_;
+    pd_key.append("@gru_pd").append(dir2str(dir)).append(std::to_string(layer));
+    auto pd = std::static_pointer_cast<dnnl::gru_forward::primitive_desc>(
+        dev_ctx_.GetBlob(pd_key));
+    if (pd == nullptr) {
+      const bool is_int8 = std::is_same<T, uint8_t>::value;
+      // Weights for int8 kernel are of a type s8
+      const auto weights_dt =
+          is_int8 ? dnnl::memory::data_type::s8 : dnnl::memory::data_type::f32;
+
+      auto x_md = MKLDNNMemDesc({Ti_, N_, ICs[layer]}, MKLDNNGetDataType<T>(),
```

Comment:
RNN primitive does support ANY memory format it would be more optimal (according to guide: https://oneapi-src.github.io/oneDNN/dev_guide_rnn.html) to apply this ANY tag for better performance.

Reply:
True, however reordering input data to the `tnc` format takes more time (double loop instead of a single one). It could be added later.

Reply:
Ok. please add a note in our internal system to track that

Reply:
Done.